File: Eventual Safe States[18-Oct-2023]/Solution.java
//   >> JAVA CODE <<

class Solution {

    public boolean dfs(int i,List<List<Integer>> adj,boolean[] visited,boolean[] rec,boolean[] presentCycle)
    {
        if(rec[i]) return presentCycle[i]=true;   
        if(visited[i]) return false;
        
        visited[i]=true;
        rec[i]=true;
        
        for(Integer in:adj.get(i))
        {
            if(dfs(in,adj,visited,rec,presentCycle)) return presentCycle[i]=true;
        }
        
        rec[i]=false;
        
            return false;
    }
    List<Integer> eventualSafeNodes(int V, List<List<Integer>> adj) {

        boolean[] visited=new boolean[V];
        boolean[] rec=new boolean[V];
        boolean[] presentCycle=new boolean[V];
        
        for(int i=0;i<V;i++)
        {
            if(visited[i]==false)
            {
                dfs(i,adj,visited,rec,presentCycle);
            }
        }
        
        List<Integer> ans=new ArrayList<>();
        
        for(int i=0;i<V;i++)
        {
            if(presentCycle[i]==false)
            {
                ans.add(i);
            }
        }

            return ans;
    }
}

File: Eventual Safe States[18-Oct-2023]/Solution.cpp
//  >> CPP CODE <<

class Solution {
  public:
    bool dfs(int i, vector<int> adj[], vector<bool>& visited, vector<bool>& rec, vector<bool>& presentCycle) {
        if (rec[i]) return presentCycle[i] = true;
        if (visited[i]) return false;

        visited[i] = true;
        rec[i] = true;

        for (int in : adj[i]) {
            if (dfs(in, adj, visited, rec, presentCycle)) return presentCycle[i] = true;
        }

        rec[i] = false;
        return false;
    }

    vector<int> eventualSafeNodes(int V, vector<int> adj[]) {
        vector<bool> visited(V, false);
        vector<bool> rec(V, false);
        vector<bool> presentCycle(V, false);

        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                dfs(i, adj, visited, rec, presentCycle);
            }
        }

        vector<int> ans;

        for (int i = 0; i < V; i++) {
            if (!presentCycle[i]) {
                ans.push_back(i);
            }
        }

        return ans;
    }
};